Networking/Networking/ApplicationPassword/RequestAuthenticator.swift

```diff
@@ -0,0 +1,91 @@
+enum RequestAuthenticatorError: Error {
+    case applicationPasswordUseCaseNotAvailable
+    case applicationPasswordNotAvailable
+}
+
+/// Authenticates request
+///
+public struct RequestAuthenticator {
+    /// Credentials.
+    ///
+    let credentials: Credentials?
+
+    /// The use case to handle authentication with application passwords.
+    ///
+    private let applicationPasswordUseCase: ApplicationPasswordUseCase?
+
+    /// Sets up the authenticator with optional credentials and application password use case.
+    /// `applicationPasswordUseCase` can be injected for unit tests.
+    ///
+    init(credentials: Credentials?, applicationPasswordUseCase: ApplicationPasswordUseCase? = nil) {
+        self.credentials = credentials
+        let useCase: ApplicationPasswordUseCase? = {
+            if let applicationPasswordUseCase {
+                return applicationPasswordUseCase
+            } else if case let .wporg(username, password, siteAddress) = credentials {
+                return try? DefaultApplicationPasswordUseCase(username: username,
+                                                              password: password,
+                                                              siteAddress: siteAddress)
+            } else {
+                return nil
+            }
+        }()
+        self.applicationPasswordUseCase = useCase
+    }
+
+    func authenticate(_ urlRequest: URLRequest) throws -> URLRequest {
+        guard case let .wporg(_, _, siteAddress) = credentials,
+              let url = urlRequest.url,
+              url.absoluteString.hasPrefix(siteAddress.trimSlashes() + "/" + RESTRequest.Settings.basePath) else {
+            // Handle non-REST requests as before
+            return try authenticateUsingWPCOMTokenIfPossible(urlRequest)
+        }
+
+        return try authenticateUsingApplicationPasswordIfPossible(urlRequest)
+    }
+
+    func generateApplicationPassword() async throws {
+        guard let applicationPasswordUseCase = applicationPasswordUseCase else {
+            throw RequestAuthenticatorError.applicationPasswordUseCaseNotAvailable
+        }
+        let _ = try await applicationPasswordUseCase.generateNewPassword()
+        return
+    }
+}
+
+private extension RequestAuthenticator {
+    /// Attempts creating a request with WPCOM token if possible.
+    ///
+    func authenticateUsingWPCOMTokenIfPossible(_ urlRequest: URLRequest) throws -> URLRequest {
+        if let credentials, case .wpcom = credentials {
+            return try AuthenticatedRequest(credentials: credentials, request: urlRequest).asURLRequest()
+        }
+        return UnauthenticatedRequest(request: urlRequest).asURLRequest()
```

Comment:
nit: it feels like this can be refactored at its call site to be like

```swift
if let credentials {
  switch credentials {
    case .wpcom:
        return try AuthenticatedRequest(credentials: credentials, request: urlRequest).asURLRequest()
    case .wporg(...):
        return ...
  }
} else {
  return UnauthenticatedRequest(request: urlRequest).asURLRequest()
}
```

Reply:
Authentication should be done after checking the contents of the `URLRequest` and not by checking the type of `Credentials`. But, I have improved this in commit 5965160b490ecd85069e2d39ab9f9c88f6cb96fa